### PR TITLE
[SPARK-11245] update twitter4j to 4.0.4 version

### PR DESCRIPTION
**Pull request by pronix**
update twitter4j to 4.0.4 version
https://issues.apache.org/jira/browse/SPARK-11245

### CODE REVIEW

**Comment by steveloughran** [comment]
Looking @ the twitter4j poms, their dependencies pull in a com.googlecode.kryo -given the kryo pain its a possible troublespot. However, that com.googlecode release is so old its in a different package from the com.esotericsoftware-packaged releases

Various modules seems to refer to a different junit (3.8.2 in http-2, 4.10 in core, ...), but those are test scoped only. Log4j 1.2.17; SLF4J 1.7.12, commons-logging 1.1 -nothing incompatible with the spark versions, afaik. 

main [pom.xml](https://github.com/yusuke/twitter4j/blob/master/twitter4j-core/pom.xml)

**Comment by pronix** [comment]
@steveloughran you mean require something change ? or just potential issue ?

**Comment by srowen** [comment]
Yeah, we would need more info here, to update across major releases. Because this is just an isolated 'leaf' package, we can probably tolerate more change there. But, at least, can you summarize any incompatible changes or siginficant ones, and more importantly key changes in the dependency tree?

**Comment by srowen** [comment]
@pronix also you need to look at https://cwiki.apache.org/confluence/display/SPARK/Contributing+to+Spark and update the title, etc.

**Comment by pronix** [comment]
@srowen Official documentation about upgrade http://blog.twitter4j.org/2014/03/13/69

**Comment by srowen** [comment]
Yes, but if you're proposing the change I believe you need to investigate the implications. This doesn't address dependency changes which Steve already alluded to.

**Comment by steveloughran** [comment]
slf4j & log4j are in sync with Spark 1.6. I can't actually see an explicit ref to commons-logging in the spark poms (just lots of exclusions of the hadoop one), but some indirect references come in at v 1.1.3.

That leaves kryo. Precisely because this is a very old import with different package names, this isn't going to be as fraught as trouble as the main kryo package.

mvn repo [says its just for testing anyway](http://mvnrepository.com/artifact/org.twitter4j/twitter4j-core/4.0.4)

No sign of the other troublespots: guava, jackson, jetty...

**Comment by srowen** [comment]
Seems pretty reasonable then. Let's see what tests say.

**Comment by pronix** [comment]
all tests are passed :)

**Comment by tdas** [comment]
There arent any real testing for twitter, as we cannot really send and receive data from twitter. So the tests passing does prove nothing.

**Comment by srowen** [comment]
Merged to master

**Comment by tdas** [comment]
@srowen Did you test it by actually passing data?

**Comment by srowen** [comment]
No, I presume that's what led @pronix to the changes here? @pronix did you try this with some real data? Yeah we don't have tests here anyway, and I'd hope that it continues to work if the API is used in the same way, but at least observing it working once is important.

**Comment by steveloughran** [comment]
Adding some tests here could be an interesting little project. the main thing to do is to have some config file in test/resources, which git is set up to ignore; if set use that to trigger the test run. We do this in hadoop for testing S3 & swift. There's still the issue that jenkins skips the work, so someone has to manually do the (slow) runs before each patch, but at least there's some automation.

**Comment by tdas** [comment]
Actually, Kinesis tests had the same problem, and i fixed it by having a
this env variable based enabling of the tests - if the env variable with
AWS configs is set then the test kick in. It would be good to add that for
twitter as well.

**Comment by steveloughran** [comment]
If you use env vars, you'd better make sure they don't ever leak into test reports (e.g no logging of env vars in an executor)

**Comment by tdas** [comment]
Yeah, gotta make sure of that. 
And i forgot to mention, those variables are set in our Jenkins configuration as secrets which gets injected as env vars automatically. So it stays safe within the realm of Jenkins admins. There is also an environmental flag called "ENABLE_KINESIS_TESTS" which need to be explicitly set to turn on the tests. The testing harness in Spark enables it only for PRs where SPARK/external/kinesis-asl is modified in the PR. Otherwise it stays unset and no tests are run, as we DO NOT WANT Kinesis tests to be running by default (cost, etc.), ONLY when the kinesis code is changed.

**Comment by steveloughran** [comment]
...so if I submit a patch which modifies that package and runs a test which prints your env vars to stdout, I've got your keys.

**Comment by tdas** [comment]
There are things like permissions and limits :)

**Comment by steveloughran** [comment]
you are right. I'd probably have to do it by releasing an upstream artifact which will wait until run on your jenkins instance before posting your credentials to a new s3n bucket which your client will also tag as world readable.
